ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,273 @@
+# Ruby SDK - Type hinting, Converters, Codecs and Errors
+
+This proposal goes into details regarding the implementation of data conversion pipeline (including
+type hints, converters and codecs) as well as outlines the hierarchy of exception classes.
+
+## Type hinting
+
+The Ruby typing system is not something we can currently rely on for specifying the types of all the
+inputs and outputs. Instead we will add custom "annotations" (technically these are just method
+calls) to the activity/workflow definitions in order to provide type hinting. Here's the basic
+example:
+
+```ruby
+class SumActivity < Temporal::Activity
+  input Integer, Integer
+  output Integer
+
+  def execute(a, b)
+    a + b
+  end
+end
+```
+
+We will support most basic Ruby types (e.g. `Integer`, `Float`, `String`, `Symbol`, `Boolean`,
+`Array`, `Hash`, etc). This will allow the SDK to perform a runtime check of the arguments and
+return values to ensure they match the specified types.
+
+We will also leverage this system to allow SDK users to specify their own types:
+
+```ruby
+class MyInput < Struct.new(:a, :b)
+  def self.load(hash)
+    new(hash['a'], hash['b'])
+  end
+
+  def dump
+    { 'a' => a, 'b' => b } # explicit for illustration purposes, can use #to_h instead
+  end
+end
+
+class MyOutput < Struct.new(:sum)
+  def self.load(sum)
+    new(sum.to_i)
+  end
+
+  def dump
+    sum.to_s # string used for illustration purposes, can use Integer directly instead
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  input MyInput
+  output MyOutput
+
+  def execute(input)
+    MyOutput.new(input.a + input.b)
+  end
+end
+```
+
+Custom types are expected to implement the following interface:
+
+```ruby
+interface _CustomType[U]
+  # Take a primitive type and return an instance of the implementing class
+  def self.load: (U) -> instance
+
+  # Returns a trivially serializable type representing the custom object
+  def dump: -> U
+end
+```
+
+The same exact concept applies to signals and queries:
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  signal :add, input: Integer
+  query :get_sum, output: MyOutput
+
+  def execute
+    ...
+  end
+end
+```
+
+Notes:
+
+- This feature is 100% optional. Without these definitions the output from converters will be used
+  directly as arguments and return values
+- The full potential of this solution is obviously limited to a single language SDKs
```

Comment:
> The full potential of this solution is obviously limited to a single language SDKs

I think I disagree with this statement and if that's the case, we should discuss in which cases this is true. JSON is the default and is supported by all SDKs and no SDK has anything language specific about JSON by default. Can you clarify why this solution is limited to this SDK? Every SDK I know of can shape a class/interface/struct to work with other languages' equivalents.

Reply:
The emphasis is on the "**full** potential". Meaning that if your Ruby workflow is expecting a non-JSON compatible type as an input, then the calling SDK (assuming non-Ruby) will have to adjust the payload to match the expectations. It is possible, but requires extra work.

Imagine you are expecting a custom class (`MyInput`) that consists of two `DateTime` values — `from` and `to`. In a Ruby-only setup the object is `dump`ed and `load`ed using the same exact class, so the experience is seamless. However in when a caller is a non-Ruby SDK — it'll have to produce a JSON that decodes into something that `MyInput.load` expects (and JSON obviously doesn't have a `DateTime` type).

So while it can work, it is not using the full potential, since either encoding or decoding will have to be tuned to match the expectations of a type hint.

Reply:
> Meaning that if your Ruby workflow is expecting a non-JSON compatible type as an input

This should not be possible unless you did a custom conversion. By default all serializable types should only support cross-SDK compatible types.

> In a Ruby-only setup the object is dumped and loaded using the same exact class, so the experience is seamless

Ah, so you did custom conversion support. w/ dump and load. Well of course custom conversion code is per SDK.

I think we're on the same page. I was just concerned that we were doing something by default that wasn't cross-SDK compatible, but I see you mean not by default.

Reply:
Yes, exactly, `dump`/`load` is for converting from the object to a converter input and vice versa. It doesn't dictate the format of the payload. A payload converter does.